uniswap-v2-core/test/UniswapV2Pair.spec.ts

```diff
@@ -437,4 +439,62 @@ describe('UniswapV2Pair', () => {
     expect(await token0.balanceOf(pair.address), "Token 0 balance of pair").to.eq(bigNumberify(1000).add(token0ExpBalVexchange))
     expect(await token1.balanceOf(pair.address), "Token 1 balance of pair").to.eq(bigNumberify(1000).add(token1ExpBalVexchange))
   })
+
+
+  /**
+   *  recoverToken - error handling for invalid tokens 
+   */
+  it('recoverToken:invalidToken', async () => {
+    await expect(pair.recoverToken(token0.address)).to.be.revertedWith('Vexchange: INVALID_TOKEN_TO_RECOVER')
+    await expect(pair.recoverToken(token1.address)).to.be.revertedWith('Vexchange: INVALID_TOKEN_TO_RECOVER')
+  
+    // address is a random invalid token address
+    await expect(pair.recoverToken("0x3704E657053C02411aA2Fd0599e75C3d817F81BC")).to.be.reverted
+  })
+
+  /**
+   *  recoverToken - failure when recoverer is AddressZero or not set
+   *  Or when the token address is not a valid ERC20 address
+   */
+  it('recoverToken:AddressZero', async () => {
+    
+    // recoverer should be AddressZero by default
+    expect(await pair.recoverer()).to.eq(AddressZero)
+    await expect(pair.recoverToken(AddressZero)).to.be.revertedWith('Vexchange: RECOVERER_ZERO_ADDRESS')
```

Comment:
Using `AddressZero` is a little ambiguous here (because we want to revert due to `recover === 0`, not invalid token. Do you want to replace with `0x3704E657053C02411aA2Fd0599e75C3d817F81BC`? Ideally store `0x3704E657053C02411aA2Fd0599e75C3d817F81BC` in a `const` within the test file

Reply:
yeah good pickup for completeness - although it will fail earlier for a separate reason, it would be better to not have an invalid input for the call we are using to test for failure of that separate reason.

There is though already a valid defined variable to use in that light: should probably just be recoverToken( token2.address ) as a valid token address?

Although redundant knowing the actual contract implementation, it may also be more complete to transfer some balance in for these types of tests are a setup step. (ie so in the end, this is testing an otherwise completely valid recover call, except for the zero recoverer address).

Reply:
yup it will be more consistent to store that invalid address in a const, so I'll do that. 

Regarding line 463, I agree that it'd be more appropriate to use a valid token address instead of AddressZero, so I'll change that. I'll also add the case of the account already having some tokens, while having recovery address zero.

Reply:
Awesome, using token2 and having token2 in the contract will prove that we are just testing the recover zero :)